File: Java/src/strategy/ducks/Duck.java
package strategy.ducks;

import strategy.behaviors.fly.FlyBehavior;
import strategy.behaviors.quack.QuackBehavior;

/**
 * This is the Duck class
 * Please put some info here.
 *
 * @author Wafer Li
 * @since 16/11/13 14:44
 */
class Duck {
    private FlyBehavior   flyBehavior;
    private QuackBehavior quackBehavior;


    Duck(FlyBehavior flyBehavior, QuackBehavior quackBehavior) {

        this.flyBehavior = flyBehavior;
        this.quackBehavior = quackBehavior;
    }


    public void display() {
        System.out.println("I am a normal duck");
    }

    public void performFly() {
        flyBehavior.fly();
    }

    public void performQuack() {
        quackBehavior.quack();
    }


    public void setFlyBehavior(FlyBehavior flyBehavior) {

        this.flyBehavior = flyBehavior;
    }


    public void setQuackBehavior(QuackBehavior quackBehavior) {

        this.quackBehavior = quackBehavior;
    }
}

File: src/kotlin_example/stategy/Duck.kt
package kotlin_example.stategy

open class Duck(var flyBehavior: FlyBehavior, var quackBehavior: QuackBehavior)
{
    open fun display() {
        println("I am a normal duck")
    }

    fun performFly() {
        flyBehavior.fly()
    }

    fun performQuack() {
        quackBehavior.quack()
    }
}

class MallardDuck(flyBehavior: FlyBehavior, quackBehavior: QuackBehavior) :
        Duck(flyBehavior, quackBehavior) {
    override fun display() {
        println("I am a mallard duck")
    }
}

class ReadHeadDuck(flyBehavior: FlyBehavior, quackBehavior: QuackBehavior) :
        Duck(flyBehavior, quackBehavior) {
    override fun display() {
        println("I am a red head duck")
    }
}

class RubberDuck(flyBehavior: FlyBehavior, quackBehavior: QuackBehavior) :
        Duck(flyBehavior, quackBehavior) {
    override fun display() {
        println("I am a rubber duck")
    }
}

class ModelDuck(flyBehavior: FlyBehavior, quackBehavior: QuackBehavior) :
        Duck(flyBehavior, quackBehavior) {
    override fun display() {
        println("I am a model duck")
    }
}
